File: misc/binarySearchTree.py
#Simple implementation of a Binary Search Tree
#Methods include insert,delete,exist,search,swap,predecessorSearch,toJson
#For the json output, you can visualize it at jsoncrack.com/editor

from random import randint

class TreeNode:
    def __init__(self,key=0,left=None,right=None):
        self.key = key
        self.left = left
        self.right = right

class BinarySearchTree:
    def __init__(self):
        self.tree = None
        self.size = 0
        self.minKey = float("inf")
        self.maxKey = float("-inf")
    
    def insert(self,key) -> None:

        def dfs(root):
            if not root:
                newNode = TreeNode(key=key,left=None,right=None)
                return newNode
            if root and key > root.key:
                root.right  = dfs(root.right)
            elif root and key < root.key:
                root.left = dfs(root.left)
            
            return root
        
        self.size += 1
        if not self.tree:
            self.tree = TreeNode(key)
        else:
            self.tree = dfs(self.tree)

        self.minKey = min(self.minKey,key)
        self.maxKey = max(self.maxKey,key)
    
    def delete(self,key) -> None:
    
        def dfs(root,key):
            if not root:
                return None
            
            if key > root.key:
                root.right = dfs(root.right,key)
            elif key < root.key:
                root.left = dfs(root.left,key)
            else: # key == root.key
                self.foundFlag = True
                #The node has only one leaf 
                if not root.right:
                    return root.left
                elif not root.left:
                    return root.right
                #The node has both leaves
                temp = root.right
                while temp.left:
                    temp = temp.left
                root.key = temp.key
                root.right = dfs(root.right,temp.key)
                    
            return root
        
        self.foundFlag = False
        if self.tree:
            self.tree = dfs(self.tree,key)
        if self.foundFlag:
            self.size -= 1
            if key == self.minKey:
                self.minKey = self.findNewMin()
            if key == self.maxKey:
                self.maxKey = self.findNewMax()
            self.foundFlag = False
        
    def exist(self,key) -> bool:
    
        def dfs(root):
        
            if not root:
                return False
            left = right = False
            if key > root.key:
                right = dfs(root.right)   
            elif key < root.key:
                left = dfs(root.left)
            else: #key == root.key:
                return True
            
            return left or right
        
        if self.tree:
            return dfs(self.tree)
        else:
            return False
    
    def search(self,key):
    
        def dfs(root):
            if not root:
                return None
            left = right = None
            
            if key > root.key:
                return dfs(root.right)
            elif key < root.key:
                return dfs(root.left)
            else:
                return root

        if self.tree:
            return dfs(self.tree)
        else:
            return None
    
    def findNewMin(self):
    
        def dfs(root):
            if not root:
                return None
            self.minimum = min(self.minimum,root.key)
            dfs(root.left)
        
        self.minimum = float("inf")
        if self.tree:
            dfs(self.tree)
        return self.minimum
    
    def findNewMax(self):
    
        def dfs(root):
            if not root:
                return None
            self.maximum = max(self.maximum,root.key)
            dfs(root.right)
        
        self.maximum = float("-inf")
        if self.tree:
            dfs(self.tree)
        return self.maximum
    
    def swap(self,oldKey,newKey) -> None:
        
        if self.exist(oldKey):
            self.delete(oldKey)
            self.insert(newKey)
    
    def predecessorSearch(self,key):
        def dfs(root,parentNode,leftParent,rightParent) -> TreeNode:
            if not root:
                return None
            
            if root.key > key:
                res = dfs(root.left,root,False,True)
                if not res and self.flag:
                    if leftParent:
                        self.flag = False
                        return parentNode
                    else:
                        return None
                elif res:
                    return res
                else:
                    return None
            elif root.key < key:
                res = dfs(root.right,root,True,False)
                if not res and self.flag:
                    if leftParent:
                        self.flag = False
                        return parentNode
                    else:
                        return None
                elif res:
                    return res
                else:
                    return None
            else: #root.key == key:
                
                #Node has left subtree
                if root.left:
                    temp = root.left
                    while temp.right:
                        temp = temp.right
                    return temp
                
                #Node has no left subtree
                else:
                    #Node is the left child of parent
                    if rightParent:
                        #Return closest ancestor which is a left parent of it's child node
                        self.flag = True
                        return None
                    #Node is the right child of parent
                    else: 
                        #Return parent node value
                        return parentNode
                    
        self.flag = False
        if self.tree and key != self.minKey and self.size > 1:
            return dfs(self.tree,None,False,False)
        else:
            return TreeNode(None)
    
    def successorSearch(self,key):
        def dfs(root,parentNode,leftParent,rightParent) -> TreeNode:
            if not root:
                return None
            
            if root.key > key:
                res = dfs(root.left,root,False,True)
                if not res and self.flag:
                    if rightParent:
                        self.flag = False
                        return parentNode
                    else:
                        return None
                elif res:
                    return res
                else:
                    return None
            elif root.key < key:
                res = dfs(root.right,root,True,False)
                if not res and self.flag:
                    if rightParent:
                        self.flag = False
                        return parentNode
                    else:
                        return None
                elif res:
                    return res
                else:
                    return None
            else: #root.key == key:
                
                #Node has right subtree
                if root.right:
                    temp = root.right
                    while temp.left:
                        temp = temp.left
                    return temp
                
                #Node has no right subtree
                else:
                    #Node is the right child of parent
                    if leftParent:
                        #Return closest ancestor which is a right parent of it's child node
                        self.flag = True
                        return None
                    #Node is the left child of parent
                    else: 
                        #Return parent node value
                        return parentNode
            
        if self.tree and key != self.minKey and self.size > 1:
            return dfs(self.tree,None,False,False)
        else:
            return TreeNode(None)
    
    def toJson(self) -> str:
        def dfs(root):
            if not root:
                return None
            #self.array.append(root.key)
            json = "{\"value\":%s,\"left\":%s,\"right\":%s}" %(root.key,dfs(root.left),dfs(root.right))
            return json
        
        if self.tree:
            json = dfs(self.tree)
            json = json.replace("None","null")
            return json
        else:
            return "{}"  
    
t = BinarySearchTree()
for i in range(20):
    t.insert(randint(0,100))
print(t.size)
print(t.minKey)
print(t.maxKey)
res = t.toJson()
print(res)

File: misc/binarySearchTree.cpp
// A simple binary search tree implementation
// Includes methods: insert,delete,exist,search,predecessor,swap,toJson

#include <iostream>
//#include <optional>
#include <string>
#include <vector> 
using namespace std;

//TreeNode class
class TreeNode {
public:
    TreeNode(int key,bool notFound,TreeNode* left,TreeNode* right);

    int key;
    bool notFound; //When a key is not found, (1) the key is set to maxKey(std::numeric_limits<int>::min() and (2) notFound is set to true
    TreeNode* left;
    TreeNode* right;
    
};

//TreeNode constructor
TreeNode::TreeNode(int key, bool notFound = false, TreeNode* left=nullptr,TreeNode* right=nullptr):key(key),notFound(notFound),left(left),right(right) {

}

//BinarySearchTree class
class BinarySearchTree {
public:
    BinarySearchTree();

    //Public Methods
    void insertNode(int key);
    void deleteNode(int key);
    bool nodeExist(int key);
    TreeNode* searchNode(int key);
    TreeNode* predecessorNode(int key);
    string toJson();
    int findNewMin();
    int findNewMax();
    void swap(int oldKey, int newKey);

    TreeNode* insertNodeDFS(TreeNode* root, int key);
    TreeNode* deleteNodeDFS(TreeNode* root, int key, bool* foundFlag);
    bool nodeExistDFS(TreeNode* root, int key);
    string toJsonDFS(TreeNode* root);
    TreeNode* searchNodeDFS(TreeNode* root, int key);
    TreeNode* predecessorNodeDFS(TreeNode* root, TreeNode* parent, int key, bool leftParent, bool rightParent, bool* searchFlag);
    void findNewMinDFS(TreeNode* root, int* minimum);
    void findNewMaxDFS(TreeNode* root, int* maximum);
    
    //Variables
    TreeNode* tree;
    int size;
    int minKey;
    int maxKey;
};

//BinarySearchTree constructor
BinarySearchTree::BinarySearchTree():tree(nullptr),size(0),minKey(std::numeric_limits<int>::max()),maxKey(std::numeric_limits<int>::min()) {

}

TreeNode* BinarySearchTree::insertNodeDFS(TreeNode* root, int key) {
    if (root==nullptr) {
        TreeNode* newNode = new TreeNode(key, false, nullptr, nullptr);
        return newNode;
    }
    if (root && key > root->key) {
        root->right = insertNodeDFS(root->right, key);
    } else if (root && key < root->key) {
        root->left = insertNodeDFS(root->left, key);
    }
    return root;
}

void BinarySearchTree::insertNode(int key) {
    size += 1;
    //If the tree has not been initialized
    if (tree==nullptr) {
        tree = new TreeNode(key,false,nullptr,nullptr);
    } else {
        tree = insertNodeDFS(tree,key);
    }
    minKey = std::min(minKey, key);
    maxKey = std::max(maxKey, key);
}

TreeNode* BinarySearchTree::deleteNodeDFS(TreeNode* root, int key, bool* foundFlag) {
    if (root==nullptr) {
        return root;
    }
    if (key > root->key) {
        root->right = deleteNodeDFS(root->right,key,foundFlag);
    } else if (key < root->key) {
        root->left = deleteNodeDFS(root->left,key,foundFlag);
    } else {
        *foundFlag = true;
        if (root->right == nullptr) {
            return root->left;
        } else if (root->left == nullptr) {
            return root->right;
        } 
        TreeNode* temp = root->right;
        while (temp->left != nullptr) {
            temp = temp->left;
        }
        root->key = temp->key;
        root->right = deleteNodeDFS(root->right, temp->key, foundFlag);
    }
    return root;
}

void BinarySearchTree::deleteNode(int key) {
    bool foundFlag = false;
    if (tree) {
        tree = deleteNodeDFS(tree, key, &foundFlag);
    }
    if (foundFlag) {
        size -= 1;
        if (key == minKey) {
            minKey = findNewMin();
        } 
        if (key == maxKey) {
            maxKey = findNewMax();
        }
        foundFlag = false;
    }
}

bool BinarySearchTree::nodeExistDFS(TreeNode* root, int key) {
    if (root == nullptr) {
        return false;
    }
    bool left = false;
    bool right = false;

    if (key > root->key) {
        right = nodeExistDFS(root->right, key);
    }
    else if (key < root->key) {
        left = nodeExistDFS(root->left, key);
    }
    else {
        return true;
    }

    return left || right;
}

bool BinarySearchTree::nodeExist(int key) {
    if (tree != nullptr) {
        return nodeExistDFS(tree, key);
    }
    else {
        return false;
    }
}

TreeNode* BinarySearchTree::searchNodeDFS(TreeNode* root, int key) {
    if (root==nullptr) {
        return root;
    } 
    TreeNode* left = nullptr;
    TreeNode* right = nullptr;

    if (key > root->key) {
        return searchNodeDFS(root->right, key);
    } else if (key < root->key) {
        return searchNodeDFS(root->left, key);
    } else {
        return root;
    }
}

TreeNode* BinarySearchTree::searchNode(int key) {
    TreeNode* notFoundObj = new TreeNode(std::numeric_limits<int>::min(),true);
    if (tree) {
        TreeNode* res = searchNodeDFS(tree,key);
        if (res == nullptr) {
            return notFoundObj;
        } else {
            return res;
        }
    } else {
        return notFoundObj;
    }
}

TreeNode* BinarySearchTree::predecessorNodeDFS(TreeNode* root, TreeNode* parent, int key, bool leftParent, bool rightParent, bool* searchFlag) {
    if (root==nullptr) {
        return root;
    }

    if (root->key > key) {
        TreeNode* res = predecessorNodeDFS(root->left, root, key, false, true, searchFlag);
        //std::optional<TreeNode*> res = predecessorNodeDFS(root->left, root, key, false, true, searchFlag);
        if (res==nullptr && searchFlag) {
            if (leftParent) {
                *searchFlag = false;
                return parent;
            } else {
                return nullptr;
            }
        } else if (res) {
            return res;
        } else {
            return nullptr;
        }
    } else if (root->key < key) {
        TreeNode* res = predecessorNodeDFS(root->right,root,key,true,false,searchFlag);
        //std::optional<TreeNode*> res = predecessorNodeDFS(root->right, root, key, true, false, searchFlag);
        if (res==nullptr && searchFlag) {
            if (leftParent) {
                *searchFlag = false;
                return parent;
            } else {
                return nullptr;
            }
        } else if (res) {
            return res;
        } else {
            return nullptr;
        }
    } else {
        if (root->left) {
            TreeNode* temp = root->left;
            while (temp->right) {
                temp = temp->right;
            }
            return temp;
        } else {
            if (rightParent) {
                *searchFlag = true;
                return nullptr;
            } else {
                return parent;
            }
        }
    }
}

TreeNode* BinarySearchTree::predecessorNode(int key) {
    bool searchFlag = false;
    TreeNode* notFoundObj = new TreeNode(std::numeric_limits<int>::min(), true);
    if (tree && key != minKey && size > 1) {
        TreeNode* res = predecessorNodeDFS(tree, NULL, key, false, false, &searchFlag);
        if (res==nullptr) {
            return notFoundObj;
        } else {
            return res;
        }
    } else {
        return notFoundObj;
    }
}

void BinarySearchTree::findNewMinDFS(TreeNode* root, int* minimum) {
    if (root==nullptr) {
        return;
    }
    *minimum = min(*minimum, root->key);
    findNewMinDFS(root->left,minimum);
}

int BinarySearchTree::findNewMin() {
    int minimum = std::numeric_limits<int>::max();
    if (tree) {
        findNewMinDFS(tree,&minimum);
    }
    return minimum;
}

void BinarySearchTree::findNewMaxDFS(TreeNode* root, int* maximum) {
    if (root == nullptr) {
        return;
    }
    *maximum = max(*maximum, root->key);
    findNewMaxDFS(root->right, maximum);
}

int BinarySearchTree::findNewMax() {
    int maximum = std::numeric_limits<int>::min();
    if (tree) {
        findNewMaxDFS(tree, &maximum);
    }
    return maximum;
}

void BinarySearchTree::swap(int oldKey, int newKey) {
    if (nodeExist(oldKey)) {
        deleteNode(oldKey);
        insertNode(newKey);
    }
}

string BinarySearchTree::toJsonDFS(TreeNode* root) {
    if (root == nullptr) {
        return "null";
    }
    string json = "{\"value\":" + std::to_string(root->key) + ", \"left\":" + toJsonDFS(root->left) + ", \"right\":" + toJsonDFS(root->right) + "}";
    return json;
}

string BinarySearchTree::toJson() {
    if (tree) {
        TreeNode* root = tree;
        return toJsonDFS(root);
    } else {
        return "{}";
    }
    
}

int main()
{
    BinarySearchTree b = BinarySearchTree();

    vector<int> nodeList = {50,17,72,12,23,54,76,9,14,19,67};
    for (int i=0; i < nodeList.size();i++) {
        b.insertNode(nodeList[i]);
    }

    printf("SearchNode 100:%d\n", b.searchNode(100)->key);
    b.swap(19, 18);
    b.swap(76, 99);
    b.swap(9, 5);
    b.deleteNode(12);
    printf("SearchNode 18:%d\n", b.searchNode(18)->key);
    printf("76 exists?%d\n", b.nodeExist(76));
    printf("99 exists?%d\n", b.nodeExist(99));
    printf("Size:%d\n", b.size);
    printf("MinKey:%d\n",b.minKey);
    printf("MaxKey:%d\n", b.maxKey);
    printf("Predecessor node of 99: %d\n", b.predecessorNode(99)->key);
    printf("Predecessor node of 12: %d\n", b.predecessorNode(12)->key);
    cout << b.toJson();
}
